ntoskrnl/io/iomgr/deviface.c

```diff
@@ -26,192 +27,542 @@ C_ASSERT(sizeof(L"{01234567-89ab-cdef-0123-456789abcdef}") == GUID_STRING_BYTES
-                               *DeviceKeyRealP,
-                               NULL);
-    Status = ZwCreateKey(InstanceKeyRealP,
-                         DesiredAccess,
-                         &ObjectAttributes,
-                         0,
-                         NULL,
-                         REG_OPTION_NON_VOLATILE,
-                         NULL);
     if (!NT_SUCCESS(Status))
     {
-        DPRINT1("Failed to open %wZ%wZ\\%wZ%\\%wZ (%x)\n", &BaseKeyU, &GuidString, &SubKeyName, &ReferenceString, Status);
-        goto cleanup;
+        DPRINT1("Failed to create or open %wZ, Status 0x%08lx\n", &ReferenceString, Status);
+        goto Quit;
     }
 
     Status = STATUS_SUCCESS;
 
-cleanup:
-    if (SubKeyName.Buffer != NULL)
-        ExFreePool(SubKeyName.Buffer);
-
+Quit:
     if (NT_SUCCESS(Status))
     {
-        if (!GuidKey)
-            ZwClose(*GuidKeyRealP);
+        if (DeviceHandle)
+            *DeviceHandle = DeviceKeyHandle;
+        else
+            ZwClose(DeviceKeyHandle);
+
+        if (DeviceDisposition)
+            *DeviceDisposition = DeviceKeyDisposition;
 
-        if (!DeviceKey)
-            ZwClose(*DeviceKeyRealP);
+        if (InstanceHandle)
+            *InstanceHandle = InstanceKeyHandle;
+        else
+            ZwClose(InstanceKeyHandle);
 
-        if (!InstanceKey)
-            ZwClose(*InstanceKeyRealP);
+        if (InstanceDisposition)
+            *InstanceDisposition = InstanceKeyDisposition;
     }
     else
     {
-        if (*GuidKeyRealP != NULL)
-            ZwClose(*GuidKeyRealP);
+        if (InstanceKeyHandle)
+            ZwClose(InstanceKeyHandle);
 
-        if (*DeviceKeyRealP != NULL)
-            ZwClose(*DeviceKeyRealP);
+        if (Create)
+            ZwDeleteKey(DeviceKeyHandle);
 
-        if (*InstanceKeyRealP != NULL)
-            ZwClose(*InstanceKeyRealP);
+        if (DeviceKeyHandle)
+            ZwClose(DeviceKeyHandle);
+
+        if (SymbolicLink.Buffer)
+            ExFreePoolWithTag(SymbolicLink.Buffer, TAG_IO);
     }
 
     return Status;
 }
 
+/**
+ * @brief
+ * Retrieves a handles to the GUID, device and instance registry keys
+ * for the specified symbolic link.
+ **/
+static
+NTSTATUS
+OpenRegistryHandlesFromSymbolicLink(
+    _In_ PCUNICODE_STRING SymbolicLinkName,
+    _In_ ACCESS_MASK DesiredAccess,
+    _Out_opt_ PHANDLE GuidKey,
+    _Out_opt_ PHANDLE DeviceKey,
+    _Out_opt_ PHANDLE InstanceKey)
+{
+    UNICODE_STRING BaseKeyU;
+    UNICODE_STRING GuidString, MungedString, ReferenceString;
+    HANDLE ClassesKey;
+    HANDLE GuidKeyReal, DeviceKeyReal, InstanceKeyReal;
+    NTSTATUS Status;
+
+    ClassesKey = GuidKeyReal = DeviceKeyReal = InstanceKeyReal = NULL;
+
+    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+
+    /* Separate symbolic link onto the parts */
+    Status = IopSeparateSymbolicLink(SymbolicLinkName,
+                                     NULL,
+                                     &MungedString,
```

Comment:
The MungedString buffer is allocated by IopSeparateSymbolicLink, but you don't free it there.

Reply:
Indeed, so I need to free it here.

Reply:
Ah, it seems to be used for DPRINT purpose only here. No any other usage of it. Then I doubt need I to receive it at all...

Reply:
indeed

Reply:
I'm freeing it properly now in the function's cleanup section, but I decided to keep it for `DPRINT` purpose, to display the detailed info about the symbolic link for which the handle(s) were not successfully opened. I can remove it still if it's important. OK?

Reply:
It's fine for me, as long as you don't leak the memory 😂

Reply:
I decided to completely remove both `MungedString` and `ReferenceString` from the function, because freeing them caused a test bots failures. `ReferenceString` had exactly the same problem with freeing it outside of `IopSeparateSymbolicLink`. I kept only `GuidString`, but that is actually used by the 2nd `IopOpenRegistryKeyEx` call below.